# Machine Learning 1 - Nearest Neighbors and Decision Trees

## Lab objectives

* Classification with decision trees and random forests.
* Cross-validation and evaluation.

In [1]:
from lab_tools import CIFAR10, get_hog_image

dataset = CIFAR10('./CIFAR10')

Pre-loading training data
Pre-loading test data


# 1. Nearest Neighbor

The following example uses the Nearest Neighbor algorithm on the Histogram of Gradient decriptors in the dataset.

In [2]:
from sklearn.neighbors import KNeighborsClassifier

clf = KNeighborsClassifier(n_neighbors=1)
clf.fit( dataset.train['hog'], dataset.train['labels'] )

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

* What is the **descriptive performance** of this classifier ?
* Modify the code to estimate the **predictive performance**.
* Use cross-validation to find the best hyper-parameters for this method.

In [7]:
from sklearn.metrics import accuracy_score

pred = clf.predict(dataset.train["hog"])
print(pred)
print(dataset.train["labels"])

score = accuracy_score(dataset.train['labels'], pred)
print(score)

[1 2 2 ... 0 0 1]
[1 2 2 ... 0 0 1]
1.0


In [9]:
import numpy as np
proportion_validation = 0.2
idxs = np.arange(len(dataset.train["labels"]))
np.random.shuffle(idxs)
print(idxs)
n_validation = int(proportion_validation * len(dataset.train["labels"]))
print(n_validation)
validation_idxs = idxs[:n_validation]
train_idxs = idxs[n_validation:]

[13000  2487 10709 ... 14025  4367  5120]
3000


In [10]:
train_X = dataset.train["hog"][train_idxs]
train_Y = dataset.train["labels"][train_idxs]

val_X = dataset.train["hog"][validation_idxs]
val_Y = dataset.train["labels"][validation_idxs]

print(train_X.shape, val_X.shape)

(12000, 256) (3000, 256)


In [15]:
clf = KNeighborsClassifier(n_neighbors=1)
clf.fit( train_X, train_Y)
pred = clf.predict(val_X)

In [16]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(val_Y, pred)
score = accuracy_score(val_Y, pred)

print(score)
print(cm)

0.701
[[631 226 144]
 [ 83 744 157]
 [ 31 256 728]]


n_neighbors=1; pas de preproc; weight uniform; dist eucl. On voit donc que le pouvoir prédictif est de 70% et le pouvoir descriptif est de 100%.

### crossvalidation

1) n_neighbors

In [19]:
from sklearn.model_selection import cross_val_score

ks = [7]

scores=[]
for k in ks:
    clf = KNeighborsClassifier(n_neighbors=k)
    scores += [cross_val_score(clf, dataset.train["hog"], dataset.train["labels"], cv=5)]

In [20]:
scores = np.array(scores)
print(scores.mean(axis=1))

[0.70793333]


n_neighbors= 1, 3, 5, 7: résultats:0.6878; 0.70606667; 0.70973333; 0.70793333. avec cv=5.
Ce paramètre est donc optimisé à 5

2) distance/metric

In [ ]:
metrics = []

scores=[]
for metric in metrics:
    clf = KNeighborsClassifier(n_neighbors=k)
    scores += [cross_val_score(clf, dataset.train["hog"], dataset.train["labels"], cv=5)]

In [ ]:
scores = np.array(scores)
print(scores.mean(axis=1))

## 2. Decision Trees

[Decision Trees](http://scikit-learn.org/stable/modules/tree.html#tree) classify the data by splitting the feature space according to simple, single-feature rules. Scikit-learn uses the [CART](https://en.wikipedia.org/wiki/Predictive_analytics#Classification_and_regression_trees_.28CART.29) algorithm for [its implementation](http://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html) of the classifier. 

* **Create a simple Decision Tree classifier** using scikit-learn and train it on the HoG training set.
* Use cross-validation to find the best hyper-paramters for this method.

In [10]:
from sklearn import tree

# --- Your code here --- #


## 3. Random Forests

[Random Forest](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) classifiers use multiple decision trees trained on "weaker" datasets (less data and/or less features), averaging the results so as to reduce over-fitting.

* Use scikit-learn to **create a Random Forest classifier** on the CIFAR data. 
* Use cross-validation to find the best hyper-paramters for this method.

In [13]:
from sklearn import ensemble

# --- Your code here --- #
